File: bin/bun-cli/src/scripts/macos/battery-info.ts
import { Script } from '../../core/decorators/Script';
import { BaseScript } from '../../core/base/Script';
import { SystemProfilerService } from '../../core/services/SystemProfilerService';
import { SystemService, PmsetBatteryData, PmsetSettings } from '../../core/services/SystemService';
import { logger } from '../../core/utils/logger';

interface BatteryData {
  systemProfiler: any;
  pmset: PmsetBatteryData;
  powerSettings: PmsetSettings;
  timestamp: string;
}

@Script({
  name: 'battery-info',
  description: 'Shows detailed battery and power charger information for macOS systems',
  emoji: '🔋',
  category: 'macos',
  arguments: '[OPTIONS]',
  options: [
    { flags: '-j, --json', description: 'Output information in JSON format' },
    { flags: '-s, --simple', description: 'Show simplified view without detailed hardware info' },
    { flags: '--no-color', description: 'Disable colored output' },
    { flags: '--refresh <seconds>', description: 'Continuously refresh every N seconds' },
  ],
  examples: [
    { command: 'battery-info', description: 'Show battery and power information' },
    { command: 'battery-info --simple', description: 'Show simplified view' },
    { command: 'battery-info --json', description: 'Output in JSON format' },
    { command: 'battery-info --refresh 30', description: 'Continuously refresh every 30 seconds' },
    { command: 'battery-info --no-color', description: 'Disable colored output' },
  ],
})
export default class BatteryInfoScript extends BaseScript {
  private jsonOutput = false;
  private simpleView = false;
  private refreshInterval?: number;
  private systemProfilerService?: SystemProfilerService;

  async run(): Promise<void> {
    // Validate macOS
    if (!SystemService.isMacOS()) {
      logger.error('This script is designed for macOS systems only');
      process.exit(1);
    }

    // Parse options
    this.jsonOutput = this.options.json || false;
    this.simpleView = this.options.simple || false;
    this.refreshInterval = this.options.refresh
      ? parseInt(this.options.refresh, 10)
      : undefined;

    // Validate refresh interval
    if (this.refreshInterval !== undefined && this.refreshInterval < 1) {
      logger.error('Refresh interval must be at least 1 second');
      process.exit(1);
    }

    // Initialize service
    this.systemProfilerService = new SystemProfilerService({
      cacheTTL: 300,
      debug: this.options.verbose,
    });

    if (this.refreshInterval) {
      logger.info(
        `Continuously refreshing every ${this.refreshInterval} seconds. Press Ctrl+C to stop.`
      );
      console.log();

      try {
        while (true) {
          SystemService.clearScreen();
          await this.displayBatteryInfo();
          await new Promise((resolve) =>
            setTimeout(resolve, this.refreshInterval! * 1000)
          );
        }
      } catch (error) {
        console.log('\n');
        logger.info('Stopped refreshing');
      }
    } else {
      logger.section('Battery & Power Information');
      await this.displayBatteryInfo();
      logger.success('Battery information display');
    }
  }

  private async displayBatteryInfo(): Promise<void> {
    const batteryData = await this.collectBatteryData();

    if (this.jsonOutput) {
      console.log(JSON.stringify(batteryData, null, 2));
      return;
    }

    this.displayOverview(batteryData);
    if (!this.simpleView) {
      this.displayBatteryDetails(batteryData);
    }
    this.displayChargerInfo(batteryData);
    if (!this.simpleView) {
      this.displayPowerSettings(batteryData);
    }
  }

  private async collectBatteryData(): Promise<BatteryData> {
    const data: BatteryData = {
      systemProfiler: {},
      pmset: {
        powerSource: 'Unknown',
        batteryId: null,
        chargePercent: 0,
        chargingState: 'Unknown',
        timeRemaining: 'Unknown',
      },
      powerSettings: {},
      timestamp: new Date().toISOString().replace('T', ' ').substring(0, 19),
    };

    // Get battery information from system profiler service
    const powerData = await this.systemProfilerService!.powerInfo(true);
    data.systemProfiler.power = powerData;

    // Get current battery status from pmset
    data.pmset = await SystemService.getPmsetBatteryStatus();

    // Get power management settings
    data.powerSettings = await SystemService.getPmsetSettings();

    return data;
  }

  private displayOverview(data: BatteryData): void {
    const pmset = data.pmset;
    const powerData = data.systemProfiler.power || {};
    const healthInfo = powerData.health_information || {};
    const chargeInfo = powerData.charge_information || {};

    console.log(`${this.batteryEmoji(pmset.chargingState)} Battery Status`);
    console.log('='.repeat(50));
    console.log();

    // Power source
    const powerSourceIcon = pmset.powerSource === 'AC Power' ? '🔌' : '🔋';
    console.log(`${powerSourceIcon} Power Source: ${pmset.powerSource}`);

    // Battery percentage with color
    const percentage = pmset.chargePercent;
    console.log(
      `${this.getPercentageEmoji(percentage)} Charge: ${percentage}% ${this.getBatteryStatusIndicator(percentage)}`
    );

    // Charging status
    const chargingIcon = this.chargingEmoji(pmset.chargingState);
    console.log(
      `${chargingIcon} Status: ${this.formatChargingState(pmset.chargingState)}`
    );

    // Time remaining
    const timeIcon = pmset.chargingState === 'charging' ? '⏱️' : '⏰';
    console.log(`${timeIcon} Time: ${pmset.timeRemaining}`);

    // Health information
    const maxCapacity = healthInfo.maximum_capacity;
    if (maxCapacity) {
      console.log(`${this.healthEmoji(maxCapacity)} Health: ${maxCapacity}`);
    }

    const cycleCount = healthInfo.cycle_count;
    if (cycleCount) {
      console.log(`${this.cycleEmoji(cycleCount)} Cycles: ${cycleCount}`);
    }

    const condition = healthInfo.condition;
    if (condition) {
      console.log(
        `${this.conditionEmoji(condition)} Condition: ${condition}`
      );
    }

    // Additional charge info
    const stateOfCharge = chargeInfo.state_of_charge;
    if (stateOfCharge !== undefined) {
      console.log(`📊 State of Charge: ${stateOfCharge}%`);
    }

    const fullyCharged = chargeInfo.fully_charged;
    if (fullyCharged !== undefined) {
      console.log(
        `${fullyCharged ? '✅' : '⏳'} Fully Charged: ${fullyCharged ? 'Yes' : 'No'}`
      );
    }

    console.log();
  }

  private displayBatteryDetails(battery: BatteryData): void {
    const powerData = battery.systemProfiler.power || {};
    const modelInfo = powerData.model_information || {};
    const chargeInfo = powerData.charge_information || {};

    logger.section('Battery Details');
    console.log();

    // Model Information
    console.log('📱 Model Information:');
    let hasModelInfo = false;

    for (const [key, value] of Object.entries(modelInfo)) {
      if (value === null || value === undefined || value === '') {
        continue;
      }

      const displayKey = key
        .split('_')
        .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
        .join(' ');
      console.log(`  ${displayKey}: ${value}`);
      hasModelInfo = true;
    }

    if (!hasModelInfo) {
      console.log('  No detailed model information available');
    }
    console.log();

    // Detailed Charge Information
    console.log('⚡ Charge Information:');
    let hasChargeInfo = false;

    for (const [key, value] of Object.entries(chargeInfo)) {
      if (value === null || value === undefined || key === 'state_of_charge') {
        continue;
      }

      const displayKey = key
        .split('_')
        .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
        .join(' ');

      if (typeof value === 'boolean') {
        console.log(`  ${displayKey}: ${value ? 'Yes' : 'No'}`);
      } else {
        console.log(`  ${displayKey}: ${value}`);
      }
      hasChargeInfo = true;
    }

    if (!hasChargeInfo) {
      console.log('  No detailed charge information available');
    }
    console.log();
  }

  private displayChargerInfo(battery: BatteryData): void {
    const powerData = battery.systemProfiler.power || {};
    const chargerInfo = powerData.ac_charger_information || {};
    const pmset = battery.pmset;

    logger.section('Charger Information');
    console.log();

    if (Object.keys(chargerInfo).length === 0) {
      console.log('ℹ️  No charger connected');
    } else {
      const connected = chargerInfo.connected === true;
      console.log(`🔌 Connected: ${connected ? 'Yes' : 'No'}`);

      if (connected) {
        console.log('⚡ Charger Specifications:');

        const wattage = chargerInfo.wattage_w;
        if (wattage) {
          console.log(`  🔋 Power: ${wattage}W`);
        }

        const id = chargerInfo.id;
        if (id) {
          console.log(`  ID: ${id}`);
        }

        const family = chargerInfo.family;
        if (family) {
          console.log(`  Family: ${family}`);
        }

        if (
          pmset.chargingState === 'charging' &&
          pmset.timeRemaining !== 'Calculating...'
        ) {
          console.log(`  ⏱️  Until Full: ${pmset.timeRemaining}`);
        }
      } else {
        console.log('ℹ️  Charger not connected or not detected');
      }
    }

    console.log();
  }

  private displayPowerSettings(battery: BatteryData): void {
    const powerData = battery.systemProfiler.power || {};
    const acPowerSettings = powerData.ac_power || {};
    const batteryPowerSettings = powerData.battery_power || {};

    logger.section('Power Management Settings');
    console.log();

    // AC Power Settings
    if (Object.keys(acPowerSettings).length > 0) {
      console.log('🔌 AC Power Settings:');
      for (const [key, value] of Object.entries(acPowerSettings)) {
        if (value === null || value === undefined) {
          continue;
        }

        const displayKey = key
          .split('_')
          .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
          .join(' ');

        if (typeof value === 'boolean') {
          console.log(`  ${displayKey}: ${value ? 'Yes' : 'No'}`);
        } else if (
          (key.includes('_minutes') || key.includes('_timer')) &&
          typeof value === 'number'
        ) {
          console.log(`  ${displayKey}: ${value} minutes`);
        } else {
          console.log(`  ${displayKey}: ${value}`);
        }
      }
      console.log();
    } else {
      console.log('ℹ️  No AC power settings available');
      console.log();
    }

    // Battery Power Settings
    if (Object.keys(batteryPowerSettings).length > 0) {
      console.log('🔋 Battery Power Settings:');
      for (const [key, value] of Object.entries(batteryPowerSettings)) {
        if (value === null || value === undefined) {
          continue;
        }

        const displayKey = key
          .split('_')
          .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
          .join(' ');

        if (typeof value === 'boolean') {
          console.log(`  ${displayKey}: ${value ? 'Yes' : 'No'}`);
        } else if (
          (key.includes('_minutes') || key.includes('_timer')) &&
          typeof value === 'number'
        ) {
          console.log(`  ${displayKey}: ${value} minutes`);
        } else {
          console.log(`  ${displayKey}: ${value}`);
        }
      }
      console.log();
    } else {
      console.log('ℹ️  No battery power settings available');
      console.log();
    }
  }

  // Helper methods for emojis and formatting
  private getBatteryStatusIndicator(percentage: number): string {
    if (percentage >= 80) return '🟢';
    if (percentage >= 50) return '🟡';
    if (percentage >= 20) return '🟠';
    return '🔴';
  }

  private getPercentageEmoji(percentage: number): string {
    if (percentage >= 10) return '🔋';
    return '🪫';
  }

  private batteryEmoji(chargingState: string): string {
    if (chargingState === 'charging' || chargingState === 'finishing charge')
      return '⚡';
    if (chargingState === 'charged') return '🔋';
    return '🪫';
  }

  private chargingEmoji(chargingState: string): string {
    switch (chargingState) {
      case 'charging':
        return '⚡';
      case 'discharging':
        return '📉';
      case 'finishing charge':
        return '🔋';
      case 'charged':
        return '✅';
      default:
        return '❓';
    }
  }

  private healthEmoji(maxCapacity: any): string {
    if (!maxCapacity) return '❓';

    let capacity: number;
    if (typeof maxCapacity === 'string') {
      const match = maxCapacity.match(/(\d+)%?/);
      capacity = match ? parseInt(match[1], 10) : parseInt(maxCapacity, 10);
    } else {
      capacity = parseInt(maxCapacity, 10);
    }

    if (capacity >= 90) return '🟢';
    if (capacity >= 80) return '🟡';
    if (capacity >= 70) return '🟠';
    return '🔴';
  }

  private conditionEmoji(condition: any): string {
    const cond = condition?.toLowerCase();
    switch (cond) {
      case 'normal':
      case 'good':
        return '🟢';
      case 'fair':
        return '🟡';
      case 'poor':
        return '🟠';
      case 'replace soon':
      case 'replace now':
      case 'service battery':
        return '🔴';
      default:
        return '❓';
    }
  }

  private cycleEmoji(cycleCount: any): string {
    if (!cycleCount) return '❓';

    let cycles: number;
    if (typeof cycleCount === 'string') {
      const match = cycleCount.match(/(\d+)/);
      cycles = match ? parseInt(match[1], 10) : parseInt(cycleCount, 10);
    } else {
      cycles = parseInt(cycleCount, 10);
    }

    if (cycles <= 300) return '🟢';
    if (cycles <= 600) return '🟡';
    if (cycles <= 1000) return '🟠';
    return '🔴';
  }

  private formatChargingState(state: string): string {
    switch (state) {
      case 'charging':
        return 'Charging ⚡';
      case 'discharging':
        return 'Discharging 📉';
      case 'finishing charge':
        return 'Finishing Charge 🔋';
      case 'charged':
        return 'Fully Charged ✅';
      default:
        return state.charAt(0).toUpperCase() + state.slice(1);
    }
  }
}

File: bin/ruby-cli/bin/battery-info.rb
#!/usr/bin/env ruby
# frozen_string_literal: true
# @category: system
# @description: Display detailed battery and power charger information for macOS
# @tags: macos, monitoring, hardware

require_relative '../../.common/script_base'
require_relative '../../.common/services/system_profiler_service'

# Script to display battery and power charger information
class BatteryInfoScript < ScriptBase
  def script_emoji; '🔋'; end
  def script_title; 'Battery & Power Info'; end
  def script_description; 'Shows detailed battery and power charger information for macOS systems'; end
  def script_arguments; '[OPTIONS]'; end

  def add_custom_options(opts)
    opts.on('-j', '--json', 'Output information in JSON format') do
      @options[:json] = true
    end
    opts.on('-s', '--simple', 'Show simplified view without detailed hardware info') do
      @options[:simple] = true
    end
    opts.on('--no-color', 'Disable colored output') do
      @options[:no_color] = true
    end
    opts.on('--refresh SECONDS', Integer, 'Continuously refresh every N seconds') do |seconds|
      @options[:refresh] = seconds
    end
  end

  def validate!
    super

    unless System.macos?
      log_error "This script is designed for macOS systems only"
      exit 1
    end

    if @options[:refresh] && @options[:refresh] < 1
      log_error "Refresh interval must be at least 1 second"
      exit 1
    end

    @json_output = @options[:json] || false
    @simple_view = @options[:simple] || false
    @refresh_interval = @options[:refresh]
  end

  def run
    if @refresh_interval
      log_info "Continuously refreshing every #{@refresh_interval} seconds. Press Ctrl+C to stop."
      puts

      begin
        loop do
          clear_screen
          display_battery_info
          sleep(@refresh_interval)
        end
      rescue Interrupt
        puts "\n"
        log_info "Stopped refreshing"
      end
    else
      log_banner("Battery & Power Information")
      display_battery_info
      show_completion("Battery information display")
    end
  end

  private

  def clear_screen
    print "\e[2J\e[H"  # ANSI escape codes to clear screen and move cursor to top
  end

  def display_battery_info
    battery_data = collect_battery_data

    if @json_output
      puts JSON.pretty_generate(battery_data)
      return
    end

    display_overview(battery_data)
    display_battery_details(battery_data) unless @simple_view
    display_charger_info(battery_data)
    display_power_settings(battery_data) unless @simple_view
  end

  def collect_battery_data
    data = {}

    # Initialize system profiler service
    @system_profiler_service ||= SystemProfilerService.new(logger: self, debug: debug?)

    # Get battery information from system profiler service
    power_data = @system_profiler_service.power_info(use_cache: true)
    data[:system_profiler] = power_data

    # Get current battery status from pmset
    data[:pmset] = parse_pmset

    # Get power management settings
    data[:power_settings] = parse_pmset_settings

    # Get current timestamp
    data[:timestamp] = Time.now.strftime("%Y-%m-%d %H:%M:%S")

    data
  end

  
  def parse_pmset
    output = System.execute("pmset -g batt", description: "Getting battery status")

    data = {
      power_source: "Unknown",
      battery_id: nil,
      charge_percent: 0,
      charging_state: "Unknown",
      time_remaining: "Unknown"
    }

    output.each_line do |line|
      next unless line.include?("Battery") || line.include?("drawing from")

      if line.include?("drawing from 'AC Power'")
        data[:power_source] = "AC Power"
      elsif line.include?("drawing from 'Battery Power'")
        data[:power_source] = "Battery Power"
      end

      # Parse battery line like: -InternalBattery-0 (id=34013283)	7%; charging; (no estimate) present: true
      if match = line.match(/-InternalBattery-\d+ \(id=(\d+)\)\s+(\d+)%;\s*(charging|discharging|finishing charge| charged);\s*(.*?)(?:\s+present: (true|false))?/)
        data[:battery_id] = match[1]
        data[:charge_percent] = match[2].to_i
        data[:charging_state] = match[3]
        data[:time_remaining] = match[4].strip.empty? ? "Calculating..." : match[4].strip
      end
    end

    data
  end

  def parse_pmset_settings
    output = System.execute("pmset -g", description: "Getting power settings")

    settings = {}
    current_power_source = nil

    output.each_line do |line|
      line = line.strip

      if line.include?("AC Power:")
        current_power_source = :ac
      elsif line.include?("Battery Power:")
        current_power_source = :battery
      elsif line.include?("Currently in use:")
        current_power_source = :current
      elsif current_power_source && line.include?(":")
        key, value = line.split(" ", 2)
        if value
          value = value.strip
          settings["#{current_power_source}_#{normalize_key(key)}"] = value
        end
      end
    end

    settings
  end

  def normalize_key(key)
    key.downcase.gsub(/[^\w]/, '_').gsub(/_+/, '_').gsub(/^_|_$/, '')
  end

  def display_overview(data)
    pmset = data[:pmset]
    power_data = data[:system_profiler]["power"] || {}
    health_info = power_data.dig("health_information") || {}
    charge_info = power_data.dig("charge_information") || {}

    puts "#{battery_emoji(pmset[:charging_state])} Battery Status"
    puts "=" * 50
    puts

    # Power source
    power_source_icon = pmset[:power_source] == "AC Power" ? "🔌" : "🔋"
    puts "#{power_source_icon} Power Source: #{pmset[:power_source]}"

    # Battery percentage with color
    percentage = pmset[:charge_percent]
    puts "#{get_percentage_emoji(percentage)} Charge: #{percentage}% #{get_battery_status_indicator(percentage)}"

    # Charging status
    charging_icon = charging_emoji(pmset[:charging_state])
    puts "#{charging_icon} Status: #{format_charging_state(pmset[:charging_state])}"

    # Time remaining
    time_icon = pmset[:charging_state] == "charging" ? "⏱️" : "⏰"
    puts "#{time_icon} Time: #{pmset[:time_remaining]}"

    # Health information from the health_information section
    max_capacity = health_info["maximum_capacity"]
    if max_capacity
      puts "#{health_emoji(max_capacity)} Health: #{max_capacity}"
    end

    cycle_count = health_info["cycle_count"]
    if cycle_count
      puts "#{cycle_emoji(cycle_count)} Cycles: #{cycle_count}"
    end

    condition = health_info["condition"]
    if condition
      puts "#{condition_emoji(condition)} Condition: #{condition}"
    end

    # Additional charge info
    state_of_charge = charge_info["state_of_charge"]
    if state_of_charge
      puts "📊 State of Charge: #{state_of_charge}%"
    end

    fully_charged = charge_info["fully_charged"]
    if fully_charged
      puts "#{fully_charged ? '✅' : '⏳'} Fully Charged: #{fully_charged ? 'Yes' : 'No'}"
    end

    puts
  end

  def display_battery_details(battery)
    power_data = battery[:system_profiler]["power"] || {}
    model_info = power_data.dig("model_information") || {}
    charge_info = power_data.dig("charge_information") || {}

    log_section "Battery Details"
    puts

    # Model Information
    puts "📱 Model Information:"
    has_model_info = false

    if model_info.any?
      model_info.each do |key, value|
        next if value.nil? || (value.is_a?(String) && value.empty?)

        # Format the key for display
        display_key = key.split('_').map(&:capitalize).join(' ')
        puts "  #{display_key}: #{value}"
        has_model_info = true
      end
    end

    puts "  No detailed model information available" unless has_model_info
    puts

    # Detailed Charge Information
    puts "⚡ Charge Information:"
    has_charge_info = false

    if charge_info.any?
      charge_info.each do |key, value|
        next if value.nil? || key == "state_of_charge" # Already shown in overview

        # Format the key for display
        display_key = key.split('_').map(&:capitalize).join(' ')

        # Special formatting for boolean values
        if value.is_a?(TrueClass) || value.is_a?(FalseClass)
          puts "  #{display_key}: #{value ? 'Yes' : 'No'}"
        else
          puts "  #{display_key}: #{value}"
        end
        has_charge_info = true
      end
    end

    puts "  No detailed charge information available" unless has_charge_info
    puts
  end

  def display_charger_info(battery)
    power_data = battery[:system_profiler]["power"] || {}
    charger_info = power_data.dig("ac_charger_information") || {}
    pmset = battery[:pmset]

    log_section "Charger Information"
    puts

    if charger_info.empty?
      puts "ℹ️  No charger connected"
    else
      # Connection status
      connected = charger_info["connected"] == true
      puts "🔌 Connected: #{connected ? 'Yes' : 'No'}"

      if connected
        # Charger specs
        puts "⚡ Charger Specifications:"

        # Display wattage prominently
        wattage = charger_info["wattage_w"]
        if wattage
          puts "  🔋 Power: #{wattage}W"
        end

        # Other specs
        id = charger_info["id"]
        puts "  ID: #{id}" if id

        family = charger_info["family"]
        puts "  Family: #{family}" if family

        # Add estimated charging time if available
        if pmset[:charging_state] == "charging" && pmset[:time_remaining] != "Calculating..."
          puts "  ⏱️  Until Full: #{pmset[:time_remaining]}"
        end
      else
        puts "ℹ️  Charger not connected or not detected"
      end
    end

    puts
  end

  def display_power_settings(battery)
    power_data = battery[:system_profiler]["power"] || {}
    ac_power_settings = power_data.dig("ac_power") || {}
    battery_power_settings = power_data.dig("battery_power") || {}

    log_section "Power Management Settings"
    puts

    # AC Power Settings
    if ac_power_settings.any?
      puts "🔌 AC Power Settings:"
      ac_power_settings.each do |key, value|
        next if value.nil?

        # Format the key for display
        display_key = key.split('_').map(&:capitalize).join(' ')

        # Special formatting for boolean values and time values
        if value.is_a?(TrueClass) || value.is_a?(FalseClass)
          puts "  #{display_key}: #{value ? 'Yes' : 'No'}"
        elsif key.include?("_minutes") && value.is_a?(Integer)
          puts "  #{display_key}: #{value} minutes"
        elsif key.include?("_timer") && value.is_a?(Integer)
          puts "  #{display_key}: #{value} minutes"
        else
          puts "  #{display_key}: #{value}"
        end
      end
      puts
    else
      puts "ℹ️  No AC power settings available"
      puts
    end

    # Battery Power Settings
    if battery_power_settings.any?
      puts "🔋 Battery Power Settings:"
      battery_power_settings.each do |key, value|
        next if value.nil?

        # Format the key for display
        display_key = key.split('_').map(&:capitalize).join(' ')

        # Special formatting for boolean values and time values
        if value.is_a?(TrueClass) || value.is_a?(FalseClass)
          puts "  #{display_key}: #{value ? 'Yes' : 'No'}"
        elsif key.include?("_minutes") && value.is_a?(Integer)
          puts "  #{display_key}: #{value} minutes"
        elsif key.include?("_timer") && value.is_a?(Integer)
          puts "  #{display_key}: #{value} minutes"
        else
          puts "  #{display_key}: #{value}"
        end
      end
      puts
    else
      puts "ℹ️  No battery power settings available"
      puts
    end
  end

  def get_battery_status_indicator(percentage)
    if percentage >= 80
      "🟢"
    elsif percentage >= 50
      "🟡"
    elsif percentage >= 20
      "🟠"
    else
      "🔴"
    end
  end

  def get_percentage_emoji(percentage)
    case percentage
    when 90..100 then "🔋"
    when 70..89 then "🔋"
    when 50..69 then "🔋"
    when 30..49 then "🔋"
    when 10..29 then "🪫"
    else "🪫"
    end
  end

  def battery_emoji(charging_state)
    case charging_state
    when "charging", "finishing charge" then "⚡"
    when "charged" then "🔋"
    else "🪫"
    end
  end

  def charging_emoji(charging_state)
    case charging_state
    when "charging" then "⚡"
    when "discharging" then "📉"
    when "finishing charge" then "🔋"
    when "charged" then "✅"
    else "❓"
    end
  end

  def health_emoji(max_capacity)
    return "❓" unless max_capacity

    # Handle both string and integer inputs
    capacity = if max_capacity.is_a?(String)
                if match = max_capacity.match(/(\d+)%/)
                  match[1].to_i
                else
                  max_capacity.to_i
                end
              elsif max_capacity.is_a?(Integer)
                max_capacity
              else
                max_capacity.to_i
              end

    case capacity
    when 90..100 then "🟢"
    when 80..89 then "🟡"
    when 70..79 then "🟠"
    else "🔴"
    end
  end

  def condition_emoji(condition)
    case condition&.downcase
    when "normal" then "🟢"
    when "good" then "🟢"
    when "fair" then "🟡"
    when "poor" then "🟠"
    when "replace soon", "replace now", "service battery" then "🔴"
    else "❓"
    end
  end

  def cycle_emoji(cycle_count)
    return "❓" unless cycle_count

    # Handle both string and integer inputs
    cycles = if cycle_count.is_a?(String)
              if match = cycle_count.match(/(\d+)/)
                match[1].to_i
              else
                cycle_count.to_i
              end
            elsif cycle_count.is_a?(Integer)
              cycle_count
            else
              cycle_count.to_i
            end

    case cycles
    when 0..300 then "🟢"
    when 301..600 then "🟡"
    when 601..1000 then "🟠"
    else "🔴"
    end
  end

  def format_charging_state(state)
    case state
    when "charging" then "Charging ⚡"
    when "discharging" then "Discharging 📉"
    when "finishing charge" then "Finishing Charge 🔋"
    when "charged" then "Fully Charged ✅"
    else state.capitalize
    end
  end

  def show_examples
    puts "Examples:"
    puts "  #{script_name}                           # Show battery and power information"
    puts "  #{script_name} --simple                  # Show simplified view"
    puts "  #{script_name} --json                    # Output in JSON format"
    puts "  #{script_name} --refresh 30              # Continuously refresh every 30 seconds"
    puts "  #{script_name} --no-color                # Disable colored output"
  end
end

BatteryInfoScript.execute if __FILE__ == $0